File: lists/arrayList/cpp/arrayList.cpp
#include <iostream>
#include "arrayList.h"

//-----------------------------------------------------------------------------

// Constructor for initializing the ArrayList
//
// If the size of the list is specified otherwise it will use the
// default size of 100.  The member access specifieds denoted m_*
// are responsible for storing the maximum ArrayList size and the
// current ArrayList size.
template <class arrayType>
ArrayList<arrayType>::ArrayList(int size) {
  std::cout << "Initializing ArrayList for data type '"
            << typeid(arrayType).name()
            << "'" << std::endl;

  m_maxSize = size;
  m_length  = 0;
  m_list    = new arrayType[size];
}

//-----------------------------------------------------------------------------

// Copy constructor for ArrayList
template <class arrayType>
ArrayList<arrayType>::ArrayList(const ArrayList<arrayType> &cpList) {
  std::cout << "Initializing ArrayList for data type '"
            << typeid(arrayType).name()
            << "' via Copy Constructor" << std::endl;

  m_maxSize = cpList.m_maxSize;
  m_length  = cpList.m_length;
  m_list    = new arrayType[m_maxSize];

  for (int i=0; i<m_length; i++) {
    m_list[i] = cpList.m_list[i];
  }
}

//-----------------------------------------------------------------------------

// Deallocate the ArrayList when done with it -- Free the memory
template <class arrayType>
ArrayList<arrayType>::~ArrayList() {
  delete[] m_list;
}

//-----------------------------------------------------------------------------

// Get the list maximum size
//
// This is simply a reference to the declared size of the array at
// initialization time
template <class arrayType>
int ArrayList<arrayType>::getMaxSize() const {
  return m_maxSize;
}

//-----------------------------------------------------------------------------

// Get the current length of the array
//
// Remember that the ArrayList is in sequential order which usese
// m_length to track where the current ArrayList content ends
template <class arrayType>
int ArrayList<arrayType>::getLength() const {
  return m_length;
}

//-----------------------------------------------------------------------------

// A check to see if the ArrayList is currently empty
template <class arrayType>
bool ArrayList<arrayType>::isEmpty() const {
  return (m_length == 0);
}

//-----------------------------------------------------------------------------

// A check to see if the ArrayList is full
//
// Cannot add additional items if the ArrayList is full
template <class arrayType>
bool ArrayList<arrayType>::isFull() const {
  return (m_length == m_maxSize);
}

//-----------------------------------------------------------------------------

// Clear the ArrayList
//
// Simply resets the ArrayList length class variable.  Doesn't
// actually remove items but just resets the m_length index
// to the beginning of the ArrayList for overwrite.
template <class arrayType>
void ArrayList<arrayType>::clear() {
  m_length = 0;
}

//-----------------------------------------------------------------------------

// Insert the item at the end of the list
//
// Will add the item at the end of the list providing the list isn't full!
template <class arrayType>
void ArrayList<arrayType>::insertAtEnd(const arrayType& value) {
  if (isFull() == 1) {
    std::cerr << "*** ERROR ***\nList is full cannot add more" << std::endl;
    return;
  } else if (searchArrayList(value) != -1) {
      std::cerr << "*** ERROR ***\n"
                   "Value '" << value << "' already exists in the list.  "
                   "Cannot have duplicates.  Not adding value!!!"
                << std::endl;
      return;
  }

  m_list[m_length] = value;

  // Increase the ArrayList end index tracker by 1
  m_length++;

}

//-----------------------------------------------------------------------------

// Insert an item at the specified index
//
// Inserts the item at the index specified so long as the index is valid and
// item does not already exist in the ArrayList
template <class arrayType>
void ArrayList<arrayType>::insertAtIndex(const int index, const arrayType& value) {
  if (index > m_maxSize) {

  } if (isFull() == 1) {
      std::cerr <<  "*** ERROR ***\nList is full cannot add more" << std::endl;
      return;
  } else if (index > m_length || index < 0) {
      std::cerr << "*** ERROR ***\n"
                   "Index is outside the range of the current list.  "
                   "Array length = " << m_length
                << " and index = "   << index
                << std::endl;
    return;
  } else if (searchArrayList(value) != -1) {
      std::cerr << "*** ERROR ***\n"
                   "Value '" << value << "' already exists in the list.  "
                   "Cannot have duplicates.  Not adding value!!!"
                << std::endl;
      return;
  }

  // Make space for the new value by moving all values to the right by one
  // for values beyond/at index
  for (int i=m_length; i > index; i--) {
    m_list[i] = m_list[i-1];
  }

  m_list[index] = value;
  m_length++;

}

//-----------------------------------------------------------------------------

// Returns whether the value at the specified index is equal to the value
// in the ArrayList
template <class arrayType>
bool ArrayList<arrayType>::isItemAtEqual(const int index, const arrayType& value) const {

  if (index > m_length || index < 0) {
    std::cerr << "*** ERROR ***\n"
                 "Index is outside the range of the current list.  "
                 "Array length = " << m_length
              << " and index = " << index
              << std::endl;
    return(0);
  }

  return(m_list[index] == value);

}

//-----------------------------------------------------------------------------

// Removes item at the specified index
//
// Removes the item at the speficied index so long as the index is valid.
// Once removed the ArrayList is shifted to keep it continuous
template <class arrayType>
void ArrayList<arrayType>::removeAtIndex(const int index) {

  if (index > m_length || index < 0) {
    std::cerr << "*** ERROR ***\n"
                 "Index is outside the range of the current list.  "
                 "Array length = " << m_length
              << " and index = " << index
              << std::endl;
    return;
  }

  // Move the items to the left to keep list continuous.
  for (int i=index; i<m_length-1; i++) {
    m_list[i] = m_list[i+1];
  }

  m_length--;

}

//-----------------------------------------------------------------------------

// Removes specified item from the ArrayList
//
// Removes an item identified if it exists in the ArrayList.  If it does not
// exist then nothing is removed.
template <class arrayType>
void ArrayList<arrayType>::remove(const arrayType& value) {

  // Get the index where the value to remove is located
  int index = searchArrayList(value);

  if (index == -1) {
    std::cerr << "*** ERROR ***\n"
                 "Value '" << value << "' is not currently in the list."
                 "Nothing to delete so doing nothing."
              << std::endl;
    return;
  }

  // Move the items to the left to keep list continuous.
  for (int i=index; i<m_length-1; i++) {
    m_list[i] = m_list[i+1];
  }

  m_length--;

}
//-----------------------------------------------------------------------------

// Get the item at the specified index
//
// If the index is not valid then return an error
template <class arrayType>
int ArrayList<arrayType>::getItemAtIndex(const int index, arrayType& value) {

  if (index < 0 || index > m_length) {
    std::cerr << "*** ERROR ***\n"
                 "Index is outside the range of the current list.  "
                 "Array length = " << m_length
              << " and index = " << index
              << std::endl;
    return -1;
  }

  value = m_list[index];
  return 0;

}

//-----------------------------------------------------------------------------

// Print the ArrayList items
//
// Prints the ArrayList at 10 items per line
template <class arrayType>
void ArrayList<arrayType>::print() const {

  std::cout << " Array List Content:" << std::endl;
  for (int i=0; i<m_length; i++) {

    if (i % 10 == 0 && i != 0) {
      std::cout << std::endl;
    }
    std::cout << m_list[i] << ", ";
  }
  std::cout << std::endl;
}

//-----------------------------------------------------------------------------

// Search the ArrayList for the given value
//
// Search the ArrayList for the given value and return the index of the match
// else return -1 to indicate that the value is not currently in the list
template <class arrayType>
int ArrayList<arrayType>::searchArrayList(const arrayType& value) const {

  for (int i=0; i<m_length; i++) {
    if (m_list[i] == value) {
      return i;
    }
  }

  return(-1);
}

//-----------------------------------------------------------------------------

int main() {

  std::cout << "****************************************************\n"
            << "*                                                  *\n"
            << "*         C++ Implementation of ArrayList          *\n"
            << "*                                                  *\n"
            << "****************************************************\n"
            << std::endl;

  // Initialize the ArrayLists
  ArrayList<int>         intList(10);
  ArrayList<double>      doubleList;
  ArrayList<std::string> stringList;

  // Create some data to populate the list
  int         intInput[10]    = {0,1,2,3,4,5,6,7,8,9};
  double      doubleInput[10] = {1.2, 2.3, 3.4, 4.5 , 5.6,
                                 6.7, 7.8, 8.9, 9.10, 10.11};
  std::string stringInput[10] = {"ab", "bc", "cd", "de", "ef",
                                 "fg", "gh", "hi", "jk", "kl"};
  for (int i=0; i<10; i++) {
    intList.insertAtEnd(intInput[i]);
    doubleList.insertAtEnd(doubleInput[i]);
    stringList.insertAtEnd(stringInput[i]);
  }

  // Print the lists
  std::cout << "\nInteger ArrayList:" << std::endl;
  intList.print();

  std::cout << "\nDouble ArrayList:" << std::endl;
  doubleList.print();

  std::cout << "\nString ArrayList:" << std::endl;
  stringList.print();

  std::cout << "\n\n ************* INTEGER ARRAYLIST TESTING ***************\n\n";

  std::cout << "\nAttempting to add '10' to integer list.  Should fail "
            << "since list is already full."
            << std::endl;
  intList.insertAtEnd(10);

  // Remove an item from the integer ArrayList
  std::cout << "\nAttempting to remove item in the third index.  "
            << "Should succeed."
            << std::endl;
  intList.removeAtIndex(3);
  std::cout << "Current List:" << std::endl;
  intList.print();

  std::cout << "\nAttempting to add '10' to integer list.  Should work "
            << "now that there is space for it."
            << std::endl;
  intList.insertAtEnd(10);
  std::cout << "Current List:" << std::endl;
  intList.print();

  std::cout << "\nAttempting to remove item label '6' from integer list.  "
            << "Should work "
            << std::endl;
  intList.remove(6);
  std::cout << "Current List:" << std::endl;
  intList.print();

  std::cout << "\n Attempting to insert an item '6' at index 2" << std::endl;
  intList.insertAtIndex(2,6);
  std::cout << "Current List:" << std::endl;
  intList.print();

  std::cout << "\nClear the list"
            << std::endl;
  intList.clear();
  std::cout << "Current List:" << std::endl;
  intList.print();

  std::cout << "\n\n ************* DOUBLE ARRAYLIST TESTING ***************\n\n";

  std::cout << "\nAttempting to add '100.9' to integer list.  Should work "
            << std::endl;
  doubleList.insertAtEnd(100.9);
  std::cout << "Current List:" << std::endl;
  doubleList.print();

  // Remove an item from the integer ArrayList
  std::cout << "\nAttempting to remove item in the 30 index.  "
            << "Should fail since nothing is there."
            << std::endl;
  doubleList.removeAtIndex(30);
  std::cout << "Current List:" << std::endl;
  doubleList.print();

  std::cout << "\nAttempting to add '1009.85' to integer list at index 9.  "
            << "Should work."
            << std::endl;
  doubleList.insertAtIndex(9, 1009.85);
  std::cout << "Current List:" << std::endl;
  doubleList.print();

  std::cout << "\nAttempting to remove item label '7.8' from integer list.  "
            << "Should work "
            << std::endl;
  doubleList.remove(7.8);
  std::cout << "Current List:" << std::endl;
  doubleList.print();

  std::cout << "\nClear the list"
            << std::endl;
  doubleList.clear();
  std::cout << "Current List:" << std::endl;
  doubleList.print();

  std::cout << "\n\n ************* STRING ARRAYLIST TESTING ***************\n\n";

  std::cout << "\nAttempting to add 'silly' to integer list.  Should work "
            << std::endl;
  stringList.insertAtEnd("silly");
  std::cout << "Current List:" << std::endl;
  stringList.print();

  // Remove an item from the integer ArrayList
  std::cout << "\nAttempting to remove item in the 30 index.  "
            << "Should fail since nothing is there."
            << std::endl;
  stringList.removeAtIndex(30);
  std::cout << "Current List:" << std::endl;
  stringList.print();

  std::cout << "\nAttempting to add 'super' to integer list at index 9.  "
            << "Should work."
            << std::endl;
  stringList.insertAtIndex(9, "super");
  std::cout << "Current List:" << std::endl;
  stringList.print();

  std::cout << "\nAttempting to remove item label 'de' from integer list.  "
            << "Should work "
            << std::endl;
  stringList.remove("de");
  std::cout << "Current List:" << std::endl;
  stringList.print();

  std::string value;
  int found;
  found = stringList.getItemAtIndex(3, value);
  if (found == -1) {
    std::cout << "\n*** ERROR ***\n"
                 "Could not get item at index 3.  Value does not exist"
              << std::endl;

  } else {
    std::cout << "\nGetting item at index 3 whose value is '"
              << value << "'" << std::endl;

  }

  std::cout << "\nClear the list"
            << std::endl;
  stringList.clear();
  std::cout << "Current List:" << std::endl;
  stringList.print();


}

File: lists/arrayList/java/ArrayList.java
public class ArrayList<T> {

  // Instance variables
  private int maxSize;
  private int length;
  private T   aList[];

  // Constructor
  public ArrayList(int size) {
    this.maxSize = size;
    this.length  = 0;
    this.aList = (T[]) new Object[size];
  }

  // --------------------------------------------------------------------------

  // Get the class instance for the maximum ArrayList size
  public int getMaxSize() {
    return this.maxSize;
  }

  // --------------------------------------------------------------------------

  // Get current length of the ArrayList
  public int getLength() {
    return this.length;
  }

  // --------------------------------------------------------------------------

  // Check if list is is full
  public boolean isFull() {
    if (this.length == this.maxSize) {
      return true;
    }
    return false;
  }

  // --------------------------------------------------------------------------

  // Check if the list is is empty
  public boolean isEmpty() {
    if (this.length == 0) {
      return true;
    }
    return false;
  }

  // --------------------------------------------------------------------------

  // Clear the list.  Simply return the length index back to zero
  public void clear() {
    this.length = 0;
  }

  // --------------------------------------------------------------------------

  // Is index valid?
  public boolean isIndexValid(int index) {
    if (index < 0 || index > this.length) {
      return false;
    }
    return true;
  }

  // --------------------------------------------------------------------------

  // Search the ArrayList for a value
  public int searchArrayList(T value) {

    // Use String class to compare so that we are comparing values
    // rather than comparing objects
    String sValue = new String();

    sValue = sValue.valueOf(value);

    for (int i=0; i<this.length; i++) {
      if (sValue.equals(String.valueOf(this.aList[i]))) {
        return i;
      }
    }
    return -1;

  }

  // --------------------------------------------------------------------------

  // Inserts an item at the end of the list if it does not already exist
  // and the list is not full
  public void insertAtEnd(T value) {

    // Is the list already full?
    if (this.isFull()) {
      System.out.println("insertAtEnd Error: List is already full.  Cannot add additional values");
      return;
    }

    // Is this a dublicate item?  No duplicates
    if (this.searchArrayList(value) != -1) {
      System.out.println("insertAtEnd Error: List is already contains value.  Cannot add duplicate values");
      return;
    }

    this.aList[this.length] = value;
    this.length++;
  }

  // --------------------------------------------------------------------------

  // Insert at a specified index.  Array cannot be full, value cannot exist
  public void insertAtIndex(int index, T value) {

    // Is the list already full?
    if (this.isFull()) {
      System.out.println("insertAtIndex Error: List is already full.  Cannot add additional values");
      return;
    }

    // Is this a dublicate item?  No duplicates
    if (this.searchArrayList(value) != -1) {
      System.out.println("insertAtIndex Error: List is already contains value.  Cannot add duplicate values");
      return;
    }

    // Make sure index range is valid
    if (this.isIndexValid(index) == false) {
      System.out.println("insertAtIndex Error: Index is out of range.  Cannot add");
    }

    // Move all items over to make room for new value
    for (int i=this.length; i>index; i--) {
      this.aList[i] = this.aList[i-1];
    }

    // Add new item
    this.aList[index] = value;
    this.length++;

  }

  // --------------------------------------------------------------------------

  // Remove an item from the list.  Item needs to be in the list to remove
  public void remove(T value) {

      int index;

      // See if value is in the ArrayList
      index = this.searchArrayList(value);

      if (index == -1) {
        System.out.println("remove Error: Value is not in list.  Cannot remove");
        return;
      }

      // Move all items to the right of removed item over
      for (int i=index; i<this.length-1; i++) {
        this.aList[i] = this.aList[i+1];
      }

      this.length--;
  }

  // --------------------------------------------------------------------------

  // Remove an item from the list for a given index.  Of course make sure the
  // index range is valid
  public void removeAtIndex(int index) {

      if (this.isIndexValid(index) == false) {
        System.out.println("removeAtIndex Error: Index range is not valid.  Cannot remove");
        return;
      }

      // Move all items to the right of removed item over
      for (int i=index; i<this.length-1; i++) {
        this.aList[i] = this.aList[i+1];
      }

      this.length--;
  }

  // --------------------------------------------------------------------------

  // Return the value at the specified index in the ArrayList.  Checks for
  // index in range of current list size.  To keep this somewhat generic
  // this returns an object and it is the responsibility of the programmer
  // to cast it approprietly
  public Object getItemAtIndex(int index) {

    Object obj = new Object();

    // Checks index is valid.  This should really be done prior to calling this
    // method since it is possible a return type of -1 could be in the list
    if (this.isIndexValid(index) == false) {
      System.out.println("getItemAtIndex Error: Index range is not valid.  Cannot remove");
      return null;
    }
    obj = this.aList[index];
    return obj;

  }

  // --------------------------------------------------------------------------

  // Prints the contents of the list
  public void printList() {

    System.out.println(" Array List Content:");
    for (int i=0; i<this.length; i++) {
      if (i%10 == 0 && i != 0) {
        System.out.println("");
      }
      if (this.aList[i] instanceof Integer) {
        System.out.printf("%d, ", this.aList[i]);
      } else if (this.aList[i] instanceof Double) {
        System.out.printf("%f, ", this.aList[i]);
      } else {
        System.out.printf("%s, ", this.aList[i]);
      }
    }
    System.out.println("");

  }

  public static void main(String[] args) {

    ArrayList<Integer> intArrayList;
    ArrayList<Double> doubleArrayList;
    ArrayList<String> strArrayList;

    // Let's test the Generic class for Integers
    int input[] = {1,2,3,4,5,6,7,8,9,10};
    double inputDouble[] = {1.1, 2.2, 3.3, 4.4, 5.5, 6.6, 7.7, 8.8, 9.9, 10.10};
    String inputStr[] = {"Hello", "There", "this", "is", "a", "String", "List"};
    int value;
    double dValue;
    String strValue;

    // Initialize the array list
    System.out.println(" TESTING INTEGER ARRAYLIST \n\n");
    intArrayList = new ArrayList<Integer>(10);

    for (int i=0; i<10; i++) {
      intArrayList.insertAtEnd(input[i]);
    }
    intArrayList.printList();

    System.out.println("\nLet's try to add '11' to the list.  Should fail, list is already full");
    intArrayList.insertAtEnd(11);
    intArrayList.printList();

    System.out.println("\nLet's remove the value '5' from the list");
    intArrayList.remove(5);
    intArrayList.printList();

    System.out.println("\nNow that we made space in the list let's add '11' to the end of the list");
    intArrayList.insertAtEnd(11);
    intArrayList.printList();

    System.out.println("\nLet's remove the value at index '0' from the list");
    intArrayList.removeAtIndex(0);
    intArrayList.printList();

    System.out.println("\nLet's remove the value at index '100' from the list.  Should fail since index is out of range");
    intArrayList.removeAtIndex(100);
    intArrayList.printList();

    System.out.println("\nLet's remove the value at the last index from the list.");
    intArrayList.removeAtIndex(8);
    intArrayList.printList();


    System.out.println("\nLet's get a value from the list at a specified index");
    System.out.println("Try to get a value from index '20' which doesn't exist so it should fail");
    if (intArrayList.isIndexValid(20)) {
      value = (Integer) intArrayList.getItemAtIndex(20);
      System.out.printf("Value got %d\n", value);
    } else {
      System.out.println("INVALID INDEX");
    }

    System.out.println("Try to get a value from index '4' which does exist so it should work");
    if (intArrayList.isIndexValid(4)) {
      value = (Integer) intArrayList.getItemAtIndex(4);
      System.out.printf("Value got %d\n", value);
    } else {
      System.out.println("INVALID INDEX");
    }

    // Initialize the array list
    System.out.println("\n\n TESTING DOUBLE ARRAYLIST \n\n");
    doubleArrayList = new ArrayList<Double>(10);

    for (int i=0; i<10; i++) {
      doubleArrayList.insertAtEnd(inputDouble[i]);
    }
    doubleArrayList.printList();

    System.out.println("\nLet's try to add '11.0' to the list.  Should fail, list is already full");
    doubleArrayList.insertAtEnd(11.0);
    doubleArrayList.printList();

    System.out.println("\nLet's remove the value '5.5' from the list");
    doubleArrayList.remove(5.5);
    doubleArrayList.printList();

    System.out.println("\nNow that we made space in the list let's add '11.0' to the end of the list");
    doubleArrayList.insertAtEnd(11.0);
    doubleArrayList.printList();

    System.out.println("\nLet's remove the value at index '0' from the list");
    doubleArrayList.removeAtIndex(0);
    doubleArrayList.printList();

    System.out.println("\nLet's remove the value at index '100' from the list.  Should fail since index is out of range");
    doubleArrayList.removeAtIndex(100);
    doubleArrayList.printList();

    System.out.println("\nLet's remove the value at the last index from the list.");
    doubleArrayList.removeAtIndex(8);
    doubleArrayList.printList();


    System.out.println("\nLet's get a value from the list at a specified index");
    System.out.println("Try to get a value from index '20' which doesn't exist so it should fail");
    if (doubleArrayList.isIndexValid(20)) {
      dValue = (Double) doubleArrayList.getItemAtIndex(20);
      System.out.printf("Value got %f\n", dValue);
    } else {
      System.out.println("INVALID INDEX");
    }

    System.out.println("Try to get a value from index '4' which does exist so it should work");
    if (doubleArrayList.isIndexValid(4)) {
      dValue = (Double) doubleArrayList.getItemAtIndex(4);
      System.out.printf("Value got %f\n", dValue);
    } else {
      System.out.println("INVALID INDEX");
    }

    // Initialize the array list
    System.out.println("\n\n TESTING STRING ARRAYLIST \n\n");
    strArrayList = new ArrayList<String>(7);

    for (int i=0; i<7; i++) {
      strArrayList.insertAtEnd(inputStr[i]);
    }
    strArrayList.printList();

    System.out.println("\nLet's try to add another item to the list.  Should fail since list is full");
    strArrayList.insertAtEnd("Super");
    strArrayList.printList();

    System.out.println("\nLet's remove 'There' so we can free some space");
    strArrayList.remove("There");
    strArrayList.printList();

    System.out.println("\nNow let's try to again to add 'Super' now that the list has space");
    strArrayList.insertAtEnd("Super");
    strArrayList.printList();

    System.out.println("\nLet's try to remove a specific index, '2'");
    strArrayList.removeAtIndex(2);
    strArrayList.printList();

    System.out.println("\nLet's try to put 'work?' into index 2");
    strArrayList.insertAtIndex(2, "work?");
    strArrayList.printList();

    System.out.println("\nLet's try to get the value at index 5");
    if (strArrayList.isIndexValid(4)) {
      strValue = (String) strArrayList.getItemAtIndex(5);
      System.out.printf("Value got '%s'\n", strValue);
    } else {
      System.out.println("INVALID INDEX");
    }

  }
}
